man/1/talc.1
.TH TALC 1
.SH NAME
talc \- a C-family statically-typed scripting language
.SH SYNOPSIS
talc [--dump-class class] [-f file] [-e program]
.SH DESCRIPTION
The language Talc is a scripting language for quick and easy programming. It is simple, straightforward, and aimed at aficionados of the C family, despite its Pascal-inspired declarations. Particularly notable features are static typing, mostly C-like syntax, and the all-important curly braces. (As Dennis Ritchie said, "C's intellectual influence on the semantic design of new languages has been small. On the other hand, it has influenced notation: even pseudo-code these days tends to contain curly braces.")

If you want a convenient language for quick little jobs, love the C family, like static typing, dislike Perl's complexity and obscurity, dislike Python's needless breaks from tradition, have been disappointed by Ruby's irregularities, and most of all don't want to throw the baby of good compile-time error checking out with the bathwater of verbosity of simple tasks, Talc may be the language for you.

Talc offers broad C, C++, and Java expression compatibility where possible, with a few little extras.
.SH FEATURES
Talc's features are best understood in comparison to other scripting languages:
.TP
.B "Run-time compiled"
Talc is a compiled language, but you don't have to recompile programs before executing them. From the user's perspective, it's interpreted, but behind the scenes, the interpreter works by compiling to JVM bytecode. This process is very fast, and the short time it takes is quickly made up by the performance of the compiled code, which is much faster than the old AST-walking interpreter.
.TP
.B "Variables have types (static typing)"
Variables (rather than just values) have associated types. Consequently, it has stronger compile time checking, and most errors are found earlier and quicker, and with less need for testing.
.TP
.B "Simple syntax"
Talc has simple syntax similar to other C-family languages, with the exception of declarations, which use a Pascal-inspired syntax.
.TP
.B "No manual memory management"
Talc has automatic memory management. Objects no longer referenced from anywhere are automatically collected by the garbage collector.
.TP
.B "Everything is an object"
Talc is a purely object-oriented language, and was ever so. Even "primitive" types such as integers are seen as objects.
.TP
.B "Class, inheritance, and methods"
As an object-oriented language, Talc has such basic features like classes, inheritance, and methods.
.TP
.B "Text processing and regular expressions"
Talc has a bunch of text processing features equivalent to those you'd find in post-Perl scripting languages.
.TP
.B "Arbitrary-precision integers"
All integer arithmetic is performed to arbitrary precision. For example, calculating 400! is trivial.
.TP
.B "Direct access to the OS"
Talc can access most Unix facilities. (FIXME: not yet fully implemented.)
.SH OPTIONS
With no arguments, interactive mode is entered. Expressions will be read from standard input, evaluated, and their results output on standard output.
.TP
\-\-\fBdump\-class\fR \fIclass\fR
Dumps a synopsis of the given \fIclass\fR to standard output.
Details include the class name, superclass, and declared member functions and member variables.
.TP
\-\-\fBdump\-classes\fR
Dumps a synopsis for every built-in class, sorted into alphabetical order.
.TP
\-\fBe\fR \fIprogram\fR
Inline mode.
The \fIprogram\fR is evaluated.
No special rules apply.
No output will be produced unless the program calls an output function.
Multiple -e arguments will be concatenated together into a single program.
Any given invocation of talc can use -e or an explicit script filename, never both.
.TP
\-\fBI\fR \fIdirectory\fR
Add \fIdirectory\fR to the library search path used by "import".
.TP
\-\fBn\fR
Adds 'while ((_ = gets()) != null) { ... }' around the user-supplied script. Note that the variable corresponding to the current line is called _ rather than $_.
.TP
\-\fBp\fR
Adds 'while ((_ = gets()) != null) { ... puts(_); }' around the user-supplied script. Note that the variable corresponding to the current line is called _ rather than $_.
.SH LEXICAL STRUCTURE
.SS Comments
Shell-like comments begin with a # and extend to the end of the line. C++-like comments begin with // and extend to the end of the line. Other than the use of # and //, shell\- and C++\-style comments are treated identically to one another.

Block comments begin with /* and extend to the next */. Block comments do not nest. Block comments make it possible to easily comment out large numbers of lines or comment out text within a line. Most scripting languages rely on dynamism and some kind of "if (false)" notation for the former use case and fail to support the latter.

.SS Keywords
The following are all keywords: "assert", "break", "class", "continue", "do", "else", "extends", "extern", "false", "final", "for", "function", "if", "implements", "import", "in", "new", "null", "return", "static", "true", "void", "while".

In the grand tradition of the C family, not all words reserved as keywords are necessarily in use. Currently "extends", "implements", and "static" are reserved but unused.
.SS Types
The basic types are "bool", "int", "object", "real", "string", and "void". The type "object" is the root of the class hierarchy. All other types have "object" as an ancestor. The type "void" is used as the return type for functions that return no value.

There are also two built-in non-scalar types: "list<T>" and "map<K,V>", representing lists whose elements are of type T and (hash) maps from key type K to value type V, respectively.
.SS Null literal
The null literal is "null".
.SS Bool literals
The bool literals are "true" and "false".
.SS Numeric literals
Various forms of numeric literal are supported. The default base is base 10, but the prefix "0b" can be used for binary literals, "0o" for octal, and "0x" for hexadecimal. Upper-case variants are not supported.

The old-fashioned interpretation of integer literals with a leading zero as octal is deliberately not implemented.

(There is currently no way to select an arbitrary base. Such a feature sounds like a nice idea, but binary is the only base most people ever need that isn't supported by traditional C-family literals, and the generality of any arbitrary-base solution is marred by the fact that it's necessary to support "0x" for the convenience of users pasting in input. Talc's scheme resembles Python, which seems the best thought-out of the existing alternatives.)

Real literals are distinguished from integer literals by the presence of a decimal point (".").

Integer literals are arbitrary precision. Real literals are currently machine precision. (In the current implementation, real literals are represented by Java's "double", which is an IEEE 754 64-bit binary floating point number.)
.SS String literals
String literals are surrounded by single quotes 'like this' or double quotes "like this". There is no difference between the two kinds of string: they are provided merely as a convenience. Use ' if your string contains ", use " if your string contains ', and use " if you've no reason to choose between the two.

The common escape characters "\\b", "\\e", "\\f", "\\n", "\\r", and "\\t" are supported and translated into ASCII backspace, ESC, form-feed, newline, carriage return, and tab respectively. The characters ", ', and \\ may also be escaped to stand for themselves. Unicode escapes of the form "\\uXXXX" where each X represents a hex digit are also supported. An attempt to escape any other character is an error.

Raw string literals are prefixed by a commercial at, @'like this' or @"like this". Backslash escape sequences are disabled in raw string literals, allowing convenient writing of otherwise awkward strings:
.nf
.sp
  path := @"c:\\windows\\paths\\";

  pattern := @"regular\\s+expressions";

  hint := @"embed double quotes like ""this""";
  easier := 'embed double quotes like "this"';
.sp
.fi
Raw string literals are of type "string", the same as normal string literals.
.SS List literals
List literals are comma-separated lists of zero or more expressions enclosed in square brackets. They have a type corresponding to a list of the type furthest from object to which all expressions are assignable:
.nf
.sp
  [ 0xcafebabe, 0xdeadbeef ]  # has type list<int>
  [ "monkey", "head" ]        # has type list<string>
  [ 1, "infinite loop" ]      # has type list<object>

  word_bag:list<string> = []; # special case, assignable to any list
.sp
.fi
As shown, the empty list is denoted [].
.SS Map literals
Map literals are comma-separated lists of one or more colon-separated key-value pairs, enclosed in square brackets. The inferred type is arrived at in the same manner as list literals, but treating the keys and values as two independent lists:
.nf
.sp
  [ 1:'one', 2:'two' ]     # has type map<int, string>
  [ '1':'one', '2':'two' ] # has type map<string, string>
  [ 1:'one', '2':'two' ]   # has type map<object, object>
  
  env:map<string, string> = [:]; # special case, assignable to any map
.sp
.fi
As shown, the empty map is denoted [:].
.SS Identifiers
Identifiers are taken from the same set as Java identifiers.
.SS Separators
The following are all separators: ":", ";", ",", ".", "(", ")", "{", "}", "[", "]".
.SS Operators
The following are all operators: "!" (prefix logical negation, postfix factorial), "-" (unary numeric negation or binary subtraction), "++", "--", "+" (numeric addition or string concatenation), "*", "**" (exponentiation), "/", "%", "<", "<=", "<<", ">", ">=", ">>", "=", "==", "!=", "&", "&&", "|", "||", "^", "~" (unary bitwise negation).

Binary operators require the types of both operands to be the same. The type of the result is the same as the type of the operands, except for relational operators, whose result is always bool. (So dividing a real by a real, for example, gives a real; dividing an int by an int gives an int, and it's not possible to divide a real by an int or an int by a real without explicitly converting one or the other to disambiguate.)

The operators "%", "<<", ">>", "&", "~", "|", "^", and "!" (postfix factorial) only operate on type int.

The operators "!" (prefix logical negation), "&&", and "||" only operate on type bool.

There are also the following compound assignment operators: "+=", "-=", "*=", "**=", "/=", "%=", "<<=", ">>=", "&=", "|=", "^=", with the usual C-family interpretation where "a op= b" is equivalent to "a = a op b".

There are also the array operators "[]" and "[]=" for array indexing and array assignment respectively. An expression such as "puts(a[i])" is translated into "puts(a.__get_item__(i))", while "a[i] = f()" is translated into "a.__set_item__(i, f())".

You can create new instances of classes with the "new" operator.

FIXME: detail operator precedence.

FIXME: explain equality/inequality.
.SH SYNTAX
.SS Blocks
A block is a sequence of statements surrounded by braces such as:
.nf
.sp
  {
    print("hello");
    print(" ");
    print("world\\n");
  }
.sp
.fi
A block is executed by executing each statement in order from first to last, though some statements may terminate the execution of the block.

Many languages treat statements and blocks interchangeably. When this documentation says "block", though, a simple statement is not acceptable. Statements such as "if" and "while", for example, always require blocks.
.SS Variable definitions
A variable definition declares one local variable and gives it an initial value:
.nf
.sp
  n: int = 0;
.sp
.fi
If the keyword "final" appears before the type, the variable may not be reassigned:
.nf
.sp
  PI: final real = 3.14;
  PI = 3.0; # compile-time error
.sp
.fi
Types are either simple types such as "object" or "int", or instantiated parametric types such as "list<string>". There is no separate array type.

As a shorthand, the type may be omitted, in which case the type is taken to be the exact type of the initializer expression. For example:
.nf
.sp
  i := 0;                 # implicitly i:int
  r := 0.0;               # implicitly r:real
  s := "hello, world!";   # implicitly s:string
  lines := s.split("\\n"); # implicitly lines:list<string>
  constant: final = 123;  # implicitly constant:final int
.sp
.fi
Note that although the ":" and "=" are two separate tokens, it's conventional not to add whitespace between them.

The type inference seen here has nothing to do with the unsafe implicit type conversions you see in other scripting languages.

Although ":=" reduces the amount of keyboarding, the program remains every bit as statically-typed as it would have been with an explicit type. It is recommended that you still use explicit types in cases where the inferred type isn't obvious, or where an explicit type seems to function as documentation. (The most important case is where the initializer is "null", but the inferred type of "object" probably wouldn't be sufficient there anyway.)

As proof that this isn't unsafe, note that the usual case where you can't use the inferred type is when you actually want a \fBless\fR specific type than would be inferred, such as "o: object = 123", which would otherwise give "o" type "int".
.SS Function definitions
A function definition looks like this:
.nf
.sp
  function int nCr(n: int, r: int) {
    return n!/(k! * (n-k)!);
  }
.sp
.fi
The keyword "function" followed by the return type and function name introduces each definition. A parenthesized comma-separated list of parameter declarations follows. Finally comes a block for the function's body.

Note that multiple parameters of the same type can be separated by commas:
.nf
.sp
  function void draw_line(x0, y0, x1, y1: int, c: color) {
    // ...
  }
.sp
.fi
Whenever you find yourself writing this, though, ask whether you'd be better served by a new type. (Here either "Point" or maybe even "Line".)
.SS Function calls
A function call looks like this:
.nf
.sp
    nCr(5, 6);
.sp
.fi
A call to a member function (also known as a method) looks like this:
.nf
.sp
    i.to_s();
.sp
.fi
Note that in the rare case where the variable "i" is replaced by a numeric literal, it is necessary to enclose the literal in parentheses or insert a space before the "." to avoid misinterpretation as a malformed real literal.
.SS Extern function definitions
An extern function definition looks like this:
.nf
.sp
  extern "Java" function real cos(a:real) = "java/lang/Math.cos:(D)D";
.sp
.fi
The string literal after extern describes the calling convention. Currently only "Java" is supported.

The format of the string literal after the assignment depends on the calling convention. In the case of Java, the string is the concatenation of the JVM-format class name, '.', the method name, ':', and the method signature.
.SS Class definitions
A class definition looks like this:
.nf
.sp
  class Point {
    x: int = 0;
    y: int = 0;

    function Point(x0: int, y0: int) {
      x = x0;
      y = y0;
    }

    function string to_s() {
      return "(" + x.to_s() + "," + y.to_s() + ")";
    }
  }
.sp
.fi
The body of a class definition contains zero or more variable definitions mixed with zero or more function definitions. A function with the same name as the class (and no return type) is the constructor, invoked by the "new" operator to initialize new instances.
.SS Empty statements
The empty statement (";") does nothing.
.SS Expression statements
An expression can be converted to a statement by following it with a semicolon (";").
.SS If statements
An "if" statement is a sequence of guard expressions with associated blocks. The first guard expression which evaluates to true will have its associated block executed. If no guard expression evaluates to true but an "else" block is present, that block will be executed instead.
.nf
.sp
  if (n == 0) {
    return "zero";
  } else if (n == 1) {
    return "one";
  } else {
    return "many";
  }
.sp
.fi
All guard expressions must be of type bool or a compile-time error results.
.SS While loops
A "while" loop executes an expression and a block repeatedly until the expression evaluates to false. The block will not be executed if the expression is false the first time it is evaluated.
.nf
.sp
  n: int = 0;
  while (n < 5) {
    puts(n);
    ++n;
  }
.sp
.fi
The expression must be of type bool or a compile-time error results.
.SS Do loops
A "do" loop executes a block and an expression repeatedly until the expression evaluates to false. The block will always be executed at least once.
.nf
.sp
  n: int = 0;
  do {
    puts(n);
    ++n;
  } while (n < 5);
.sp
.fi
The expression must be of type bool or a compile-time error results.
.SS For loops
A "for" loop initializes a variable local to the statement before executing a continuation expression, a block, and an update expression until the continuation expression evaluates to false. The continuation expression must be of type bool or a compile-time error results. The update expression can be of any type.
.nf
.sp
  for (n: int = 0; n < 5; ++n) {
    puts(n);
  }
.sp
.fi
Note that Talc's for loop is slightly more restricted than in most C-family languages, because Talc has no comma operator.
.SS For-each loops
A for-each loop iterates over a collection, evaluating a block once for each value in the collection. For example:
.nf
.sp
  # Iterate over the values in the collection:
  for (word in [ "hello", "world" ]) {
    puts(word);
  }

  # Iterate over the key, value pairs in the collection:
  for (k, v in [ 123:"hello", 456:"world" ]) {
    puts(k, " : ", v);
  }
  # (For a list, the keys are the integer indexes.)
  for (i, word in [ "hello", "world" ]) {
    puts(i, " : ", word);
  }
  
  # Iterate over the characters in a string:
  for (c in "hello, world\n") {
    print(c);
  }
.sp
.fi
The expression must be of list, map, or string type. (It will be extended to user-defined types eventually, but that's not been implemented yet.) The for-each loop has one or two loop variable names declared. If it has one, it is given the value type of the collection. If it has two, the first is given the key type of the collection and the second the value type. It is not possible to specify explicit types, nor to reverse the order of declaration.

A more Java-like syntax using a ":" instead of the "in" keyword isn't possible because of the ambiguity with an explicit type declaration. Talc initially used ";" (following the D language) but it looked strange, looked overly similar to the normal for loop, and was surprisingly hard to remember. A "foreach" keyword was considered (both in conjunction with ":" and "in"), but rejected as not being a real word. Though in some ways that's an ideal property for something as disruptive as a keyword.
.SS Break statements
A "break" statement transfers control out of the innermost enclosing "do", "for", or "while" statement.
.SS Continue statements
A "continue" statement transfers control to the loop-continuation test of the innermost enclosing "do", "for", or "while" statement.
.SS Return statements
A "return" statement returns control to the invoker of a function.

If the enclosing function has return type "void", supplying a non-void expression to the "return" statement will result in a compile-time error, but an expression of void type, such as a call to a void function, is fine.

For non-void functions, omitting an expression or providing an expression of an inappropriate type will result in compile-time errors.
.SS Assert statements
An "assert" statement consists of a boolean expression optionally followed by an explanatory expression of any type. If the boolean expression evaluates to false, an exception is thrown. If an explanatory expression was provided, it will be evaluated and used as the exception's detail message. If the boolean expression evaluates to true, the "assert" statement completes normally.
.nf
.sp
  assert list.size() == 0 : "list not empty!";
.sp
.fi
Unlike in other languages, assertions cannot be turned off at either compile-time or run-time.

src/org/jessies/talc/Talc.java
/*
 * This file is part of Talc.
 * Copyright (C) 2007-2009 Elliott Hughes <enh@jessies.org>.
 * 
 * Talc is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 * 
 * Talc is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.jessies.talc;

import java.io.*;
import java.util.*;

public class Talc {
    private static final boolean[] debuggingFlags = new boolean[127];
    private static final String[] debuggingFlagNames = new String[127];
    static {
        debuggingFlagNames['C'] = "don't use a synthetic 'constant pool' for int and real constants";
        debuggingFlagNames['i'] = "show each inferred type as it's fixed up";
        debuggingFlagNames['l'] = "show each token returned by the lexer";
        debuggingFlagNames['n'] = "don't execute the generated code";
        debuggingFlagNames['o'] = "don't optimize the AST before generating code";
        debuggingFlagNames['p'] = "show information about parsing as it progresses, and the AST for each completed parse";
        debuggingFlagNames['t'] = "show timing information for each phase of compilation/execution";
        debuggingFlagNames['T'] = "show information helpful when debugging the type checker";
        debuggingFlagNames['S'] = "show the generated JVM bytecodes";
        debuggingFlagNames['s'] = "save the generated code to /tmp";
        debuggingFlagNames['v'] = "verify the generated code with ASM's verifier (implies 's'; libasm3-java must be installed)";
    }
    
    private ArrayList<String> libraryPath;
    
    private boolean implicitInputMode = false; // -n
    private boolean implicitInputOutputMode = false; // -p
    
    public Talc() {
        initLibraryPath();
    }
    
    private void initLibraryPath() {
        libraryPath = new ArrayList<String>();
        libraryPath.add("/usr/lib/talc/");
        libraryPath.add(System.getProperty("org.jessies.projectRoot") + "/lib/talc/");
    }
    
    private void reportTime(String task, long ns) {
        if (Talc.debugging('t')) {
            double s = ns/1000000000.0;
            System.err.println("[talc] " + new java.text.DecimalFormat("#.####").format(s) + "s " + task);
            // GNU Classpath's String.format seems to be broken at the moment, preventing me from using this:
            //System.err.println(String.format("[talc] %.4fs %s", s, task));
        }
    }
    
    private void parseAndEvaluate(String argv0, String[] args, Lexer lexer) throws Throwable {
        // 1. Parse.
        long parse0 = System.nanoTime();
        List<AstNode> ast = new Parser(lexer, libraryPath).parse();
        reportTime("parsing", System.nanoTime() - parse0);
        // 1a. Add any code implied by -n or -p options.
        ast = addImplicitCode(ast);
        // 1b. Insert values for built-in constants.
        Scope.initGlobalScope(argv0);
        
        // 2. Compile-time checking.
        // 2a. Set up the symbol table.
        SymbolTable symbolTable = new SymbolTable(ast);
        reportTime("symbol table construction", System.nanoTime() - symbolTable.creationTime());
        // 2b. Type checking.
        AstTypeChecker typeChecker = new AstTypeChecker(ast);
        reportTime("type checking", System.nanoTime() - typeChecker.creationTime());
        // 2c. Non-type checks (which can assume that type-checking passed).
        AstErrorChecker errorChecker = new AstErrorChecker(ast);
        reportTime("other checking", System.nanoTime() - errorChecker.creationTime());
        // 2d. Simplification.
        if (Talc.debugging('o') == false) {
            AstSimplifier simplifier = new AstSimplifier();
            ast = simplifier.simplify(ast);
            reportTime("simplification", System.nanoTime() - simplifier.creationTime());
        }
        
        // 3. Byte-code generation.
        TalcClassLoader loader = new TalcClassLoader();
        JvmCodeGenerator codeGenerator = new JvmCodeGenerator(loader, ast);
        reportTime("code generation", System.nanoTime() - codeGenerator.creationTime());
        
        // 4. Execution.
        if (Talc.debugging('n')) {
            System.err.println("[talc] (not executing generated code because of -D n.)");
            return;
        }
        long execution0 = System.nanoTime();
        try {
            Class<?> generatedClass = loader.getClass("GeneratedClass");
            generatedClass.getMethod("main", String[].class).invoke(null, (Object) args);
        } finally {
            reportTime("execution", System.nanoTime() - execution0);
        }
    }
    
    private List<AstNode> addImplicitCode(List<AstNode> ast) {
        if (!implicitInputMode && !implicitInputOutputMode) {
            return ast;
        }
        
        // _: string = null; while ((_ = gets()) != null) { ... puts(_); }
        AstNode _ = new AstNode.VariableName(null, "_");
        AstNode nullLiteral = new AstNode.Constant(SourceLocation.NONE, null, TalcType.NULL);
        AstNode lhs = new AstNode.BinaryOperator(SourceLocation.NONE, Token.ASSIGN, _, new AstNode.FunctionCall(SourceLocation.NONE, "gets", (AstNode) null, new AstNode[0]));
        AstNode expression = new AstNode.BinaryOperator(SourceLocation.NONE, Token.NE, lhs, nullLiteral);
        
        ArrayList<AstNode> loopBody = new ArrayList<AstNode>();
        loopBody.addAll(ast);
        if (implicitInputOutputMode) {
            loopBody.add(new AstNode.FunctionCall(SourceLocation.NONE, "puts", (AstNode) null, new AstNode[] { _ }));
        }
        
        ArrayList<AstNode> result = new ArrayList<AstNode>();
        result.add(new AstNode.VariableDefinition(SourceLocation.NONE, "_", TalcType.STRING, nullLiteral, false));
        result.add(new AstNode.WhileStatement(SourceLocation.NONE, expression, new AstNode.Block(SourceLocation.NONE, loopBody)));
        return result;
    }
    
    private void interactiveReadEvaluatePrintLoop() throws IOException {
        LineReader lineReader = new LineReader();
        String line;
        // FIXME: is breaking into lines really useful, when there's no context carried over?
        while ((line = lineReader.readLine("  ")) != null) {
            try {
                if (line.endsWith(";") == false && line.endsWith("}") == false) {
                    // For interactive use, supply a trailing semicolon so the user isn't forced to type it.
                    // FIXME: is this really useful?
                    line += ";";
                }
                parseAndEvaluate(null, new String[0], new Lexer(line));
            } catch (Throwable th) {
                reportError(th);
            }
        }
        System.out.print("\b\b");
    }
    
    private void usage(int exitStatus) {
        PrintStream out = (exitStatus == 0) ? System.out : System.err;
        out.println("usage: talc [talc-arguments] [--] [script-filename] [script-arguments]");
        out.println("  -D flags           set debugging flags (-D ? for a list)");
        out.println("  --dump-class name  describe the given class");
        out.println("  --dump-classes     describe all built-in classes");
        out.println("  -e program         one line of program; multiple -e's allowed, but omit explicit script filename");
        out.println("  -I directory       add the given directory to the \"import\" search path");
        out.println("  -n                 assume 'while ((_ = gets()) != null) { ... }' around script");
        out.println("  -p                 assume 'while ((_ = gets()) != null) { ... puts(_); }' around script");
        out.println("  --copyright        show brief copyright information");
        System.exit(exitStatus);
    }
    
    private void parseArguments(String[] args) throws Throwable {
        String scriptFilename = null;
        String expression = null;
        ArrayList<String> scriptArgs = new ArrayList<String>();
        boolean inScriptArgs = false;
        boolean didSomethingUseful = false;
        // FIXME: this should be more like getopt(3) to cope with stuff like "-ne 'puts(_.uc());'"
        for (int i = 0; i < args.length; ++i) {
            if (inScriptArgs) {
                scriptArgs.add(args[i]);
            } else if (args[i].equals("--copyright")) {
                reportCopyright();
                didSomethingUseful = true;
            } else if (args[i].equals("-h") || args[i].equals("--help")) {
                usage(0);
            } else if (args[i].startsWith("-D")) {
                String flags = args[i].substring(2);
                if (flags.length() == 0) {
                    if (i + 1 >= args.length) {
                        usage(1);
                    }
                    flags = args[++i];
                }
                parseDebuggingFlags(flags);
            } else if (args[i].equals("--dump-class")) {
                if (i + 1 >= args.length) {
                    usage(1);
                }
                String typeName = args[++i];
                TalcType type = TalcType.byName(typeName);
                if (type == null) {
                    die("unknown type \"" + typeName + "\"");
                } else {
                    System.out.print(type.describeClass());
                }
                didSomethingUseful = true;
            } else if (args[i].equals("--dump-classes")) {
                for (TalcType type : TalcType.documentedTypes()) {
                    System.out.println(type.describeClass());
                    System.out.println();
                }
                didSomethingUseful = true;
            } else if (args[i].equals("-e")) {
                if (i + 1 >= args.length) {
                    usage(1);
                }
                String newExpression = args[++i];
                if (expression == null) {
                    expression = newExpression;
                } else {
                    // Multiple -e's are allowed, and concatenated.
                    expression += "\n" + newExpression;
                }
            } else if (args[i].startsWith("-I")) {
                String directory = args[i].substring(2);
                if (directory.length() == 0) {
                    if (i + 1 >= args.length) {
                        usage(1);
                    }
                    directory = args[++i];
                }
                libraryPath.add(directory);
            } else if (args[i].equals("-n")) {
                implicitInputMode = true;
            } else if (args[i].equals("-p")) {
                implicitInputOutputMode = true;
            } else if (args[i].equals("--")) {
                inScriptArgs = true;
            } else if (args[i].startsWith("-")) {
                die("unrecognized option \"" + args[i] + "\"");
            } else {
                scriptFilename = args[i];
                inScriptArgs = true;
            }
        }
        if (scriptFilename == null && expression == null) {
            if (didSomethingUseful) {
                // Fair enough, then.
                System.exit(0);
            } else {
                die("no script filename supplied");
            }
        }
        if (scriptFilename != null && expression != null) {
            die("can't mix -e and an explicit script filename (\"" + scriptFilename + "\")");
        }
        if (implicitInputMode && implicitInputOutputMode) {
            die("can't use -n and -p together");
        }
        Lexer lexer = (scriptFilename != null) ? new Lexer(new File(scriptFilename)) : new Lexer(expression);
        parseAndEvaluate(scriptFilename, scriptArgs.toArray(new String[scriptArgs.size()]), lexer);
    }
    
    private static void die(String message) {
        System.err.println("talc: error: " + message);
        System.exit(1);
    }
    
    private static void reportError(Throwable th) {
        if (th instanceof TalcError) {
            System.err.println(th.getMessage());
            System.err.println("Java stack:");
            for (StackTraceElement e : th.getStackTrace()) {
                System.err.println("    " + e);
            }
        } else if (th instanceof java.lang.reflect.InvocationTargetException) {
            // The compiled code threw an exception, and Java wrapped it.
            // Pull out and report the original exception.
            // FIXME: hide the part of the stack that isn't part of the user's code?
            th.getCause().printStackTrace();
        } else {
            System.err.println("Unexpected internal error:");
            th.printStackTrace();
        }
    }
    
    private static void reportCopyright() {
        PrintStream out = System.out;
        out.println("talc - http://code.google.com/p/talc/");
        out.println();
        out.println("Copyright (C) 2007-2009 Elliott Hughes <enh@jessies.org>.");
        out.println();
        out.println("Talc uses org.mozilla.classfile, written by Roger Lawrence.");
        out.println();
        out.println("Talc is free software; you can redistribute it and/or modify");
        out.println("it under the terms of the GNU General Public License as published by");
        out.println("the Free Software Foundation; either version 3 of the License, or");
        out.println("(at your option) any later version.");
        out.println();
        out.println("Talc is distributed in the hope that it will be useful,");
        out.println("but WITHOUT ANY WARRANTY; without even the implied warranty of");
        out.println("MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the");
        out.println("GNU General Public License for more details.");
        out.println();
        out.println("You should have received a copy of the GNU General Public License");
        out.println("along with this program.  If not, see <http://www.gnu.org/licenses/>.");
        out.println();
        out.println("Documentation should be available via \"man talc\" or on the web.");
    }
    
    public static boolean debugging(char ch) {
        if (ch > debuggingFlags.length || debuggingFlagNames[ch] == null) {
            throw new RuntimeException("unknown debugging flag '" + Lexer.ensurePrintable(ch) + "'");
        }
        return debuggingFlags[ch];
    }
    
    private void parseDebuggingFlags(String flags) {
        for (int i = 0; i < flags.length(); ++i) {
            char ch = flags.charAt(i);
            if (debuggingFlagNames[ch] == null) {
                reportUnknownDebuggingFlag(ch);
            } else {
                debuggingFlags[ch] = true;
            }
        }
    }
    
    private void reportUnknownDebuggingFlag(char unknownFlag) {
        System.err.println("talc: unknown debugging flag '" + unknownFlag + "'; valid flags are:");
        for (int i = 0; i < debuggingFlagNames.length; ++i) {
            if (debuggingFlagNames[i] != null) {
                System.err.println(" " + ((char) i) + " - " + debuggingFlagNames[i]);
            }
        }
        System.exit(1);
    }
    
    public static void main(String[] args) throws IOException {
        Talc talc = new Talc();
        if (args.length == 0) {
            talc.interactiveReadEvaluatePrintLoop();
        } else {
            try {
                talc.parseArguments(args);
            } catch (Throwable th) {
                reportError(th);
                System.exit(1);
            }
        }
    }
}
